### ISSUE TITLE
hot reloading doesn't work outside a package with type:module

### ISSUE THREAD

**Issue by pmcelhaney** [pmcelhaney/counterfact#765]
Counterfact transpiles TypeScript files to JS and puts them in a folder called `.cache`. Those files are `import()`ed with a query string at the end of the URL to bust the cache. The cache busing mechanism doesn't work if the files are CJS.

To ensure the files are modules, we used the `.mjs` extension on transpiled files. That worked great until I wanted to import another file from my `_.context.ts` file. The import didn't work because it was trying to import a `.js` file instead of a `.mjs` file. I could probably mess around with the TypeScript settings to make it work, but that felt clunky.

So I went back to transpiling to `.js` files and looked for a package.json to determine whether those files should be modules or CommonJS. That worked great except for the fact that hot reloading doesn't work in CommonJS.

A few possible ways to solve this problem:

1. Add a package.json to the .cache directory. The downside is if you want the generated code to be inside an existing package, it will create a new package which is not desirable. We could mitigate by creating the package.json if and only if there isn't already a package.json higher up with `"type": "module"`. That would work unless someone wants to use Counterfact inside a CommonJS module. At this point, I'm not particularly concerned about supporting that use case. 

2. Load modules using a VM. This might be the simplest and best option. Since you pass code to a VM instead of a file name, we can be sure that caching is not an issue. We can also ensure that files that are imported / required are not cached. A whole other problem that other solutions don't address.

3. Go back to `.mjs` files.

4. <del>Use `module.require()` instead of `import()` and manually manage Node's cache?</del> that doesn't work  

Okay, now that I've written that out I think option 2 is going to make the most sense, with option 1 as a stopgap. I did a little bit of testing and it looks like we can use vm.Script() even when loading modules. We don't need to use the experimental VM modules. For CJS we can pass a modified version of `require()` to the context that busts cache. I'm not sure how `import()` would work but I guess it's a similar idea.

**Comment by dethell** [comment]
Option 3 isn't great because of the import issue. We want to be able to import `.js` files as desired. Using a VM seems cleanest and would also clean up the need to even have a `.cache` folder. Is there a concern about performance running that way? I'm guessing this is similar to running code in webpack dev server in terms of performance?

**Comment by pmcelhaney** [comment]
I found another option. 

If we detect that CommonJS is being used, instead of `import()` we can can call a `cjsImportWithNoCache()` function.

That function will be defined inside a `.cjs` file. Because the file is CJS, we can use `require()` and we can invalidate the cache by deleting the key under `require.cache`.

It's kludgy, but I've tested on Node 18 and 21 and it works. 

A VM would be better, but at this point I'd rather not rely on the --experimental-vm-modules flag.

**Comment by pmcelhaney** [comment]
Currently we have this code:

```js
  const moduleKind = await determineModuleKind(compiledPathsDirectory);

  const transpiler = new Transpiler(
    nodePath.join(modulesPath, "paths").replaceAll("\\", "/"),
    compiledPathsDirectory,
    moduleKind,
  );

  const moduleLoader = new ModuleLoader(
    compiledPathsDirectory,
    registry,
    contextRegistry,
  );

```

The first step I think is to refactor so that we pass the source directory as the first argument to ModuleLoader, and have ModuleLoader create the Transpiler. That would make `compiledPathsDirectory` (i.e. the `.cache` directory) an implementation detail of ModuleLoader.

`determineModuleKind()` can also get stuffed inside of `ModuleLoader`.

Then we can factor out an `uncachedImport()` function with one of two implementations, depending on the module kind:

- `uncached-import.js` uses `import()` and attaches a random querystring to the end of the module name.
- `uncached-import.cjs` uses `require()` after deleting the `require.cache` entry.

That also paves the way for eventually getting rid of the `.cache` directory and using a VM.

**Comment by pmcelhaney** [comment]
Fixed in 0.37.1

I used the trick described in my last comment.
